Add OPA policy for SPIRE API authorization

lumjjb opened the pull request:
<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/main/CONTRIBUTING.md

2. Please remember to include a DCO on every commit (`git commit -s`)
https://github.com/apps/dco
-->

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md? 🤞 
- [x] Proper tests/regressions included?
- [ ] Documentation updated? Want some feedback on this!
 
**Affected functionality**
<!-- Please provide a description of the affected functionality -->
Changes SPIRE authorization mechanics to use OPA instead. Allows extensibility of authorization policies to be written in rego, allowing more granular policy decisions. This can be done by adding the appropriate rego files and JSON data via SPIRE server config. If unconfigured, defaults to the standard SPIRE authorization policy.

**Description of change**
<!-- Please provide a description of the change -->
Changed the middleware authorization preprocessing to call out to OPA instead, part of this makes the old authorization code obsolete since it calls out to OPA for the decisions, thus removes the old authz code. The main changes of this PR are the server authorization middleware, the server endpoint as well as the `common/policy`

**Which issue this PR fixes**
<!-- optional. `fixes #<issue number>` format will close an issue when this PR is merged -->
This fixes https://github.com/spiffe/spire/issues/1975

lumjjb commented:
@APTy can I add the DCO for your commit on your behalf? :)

azdagron commented:
Looks like there are bugs in the downstream authorization logic.

> 2021/07/13 16:47:58 RPC "NewDownstreamX509CA": failed to call NewDownstreamX509CA: rpc error: code = Internal desc = caller is not a downstream workload

lumjjb commented:
Thanks @azdagron , i think i see what the issue is now... that error message is from https://github.com/spiffe/spire/blob/main/pkg/server/api/svid/v1/service.go#L321-L324

Which seems to gather the downstream entries from the context... I think this was implicitly done in the old authorization code, I just added code to do that in the authorization. So 🤞 it solves it..

I have some thoughts on this, I'd like to get your feedback... but all condition on whether this is the actual cause of the failure.... (so we can wait for the CI first before discussing further, if not we can ignore the following :D, and probably comes back to me going back to the drawing board...)

However, I do have a question on that, if this is the explicit behavior of the authorization middleware, then perhaps I should add some tests to check for this as well... However, this gets a bit puzzling because there is a reliance on the call to authorization... Which in the case of an authorization policy of "allow all", would not result in this context being set... This is true of the old implementation as well. Which makes me think whether this is something that should be addressed more generally as a preprocessor? or not explicitly relying on `rpccontext.CallerDownstreamEntries(ctx)`, and checking if it doesn't have any entries explicitly set...

lumjjb commented:
Updated with docs

azdagron commented on pkg/common/policy/policy_examples.go:
```diff
@@ -0,0 +1,349 @@
+package policy
+
+const defaultPermissionData = `
```
It's a little weird for server-specific defaults to live under `common`. I wonder if there shouldn't be a `pkg/server/policy` package that has a `NewEngineOrDefault()` method that either constructs an engine with the provided config or returns an engine with the default policy loaded. That would also get rid of my concern about NewEngine returning `nil`.

azdagron commented on pkg/common/policy/policy_examples.go:
```diff
@@ -0,0 +1,349 @@
+
+const defaultPermissionData = `
+{
+    "apis": [
```
These defaults are now listed in three spots (MD, conf/server/permissions.json, and here). I think we should try and consolidate them into one spot. One solution is to create the `pkg/server/policy` package, include the default files in that location, and embed them into the binary using the `go:embed` directive (https://pkg.go.dev/embed). The markdown could then reference that file directly and we only have one spot to update things.

As I mentioned elsewhere, I don't think the "developer" conf under conf/ should configure the policy engine since most users will probably use the default.

lumjjb replied:
I am all for using embed.. is there any concern with the toolchain, i think `embed` is a fairly new golang feature - just checking if this is an issue for broader use cases or CI

>As I mentioned elsewhere, I don't think the "developer" conf under conf/ should configure the policy engine since most users will probably use the default.

yea agree - can change this to not set it and let it use the default.

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+	}
+
+	exp := rs[0].Expressions[0]
+	resultMap := exp.Value.(map[string]interface{})
```
Is the value guaranteed to be of this type? If not, we should use the comma-ok pattern on the type coercion to prevent a panic and return an error instead.

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+	resultMap := exp.Value.(map[string]interface{})
+
+	var ok bool
+	result.Allow, ok = resultMap["allow"].(bool)
```
nit: `"allow"` and friends should probably be constants.

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+		return result, errors.New("policy: result did not contain \"allow\" bool value")
+	}
+
+	result.AllowIfAdmin, ok = resultMap["allow_if_admin"].(bool)
+	if !ok {
+		return result, errors.New("policy: result did not contain \"allow_if_admin\" bool value")
+	}
```
This repetition could be cleaned up by defining a function-local helper:
```suggestion
	getBoolValue := func(name string) (bool, error) {
		value, ok := resultMap[name].(bool)
		if !ok {
			return false, fmt.Errorf("policy: result did not contain %q bool value", name)
		}
		return value, nil
	}

	if result.AllowIfAdmin, err = getBoolValue(allowIfAdminKey); err != nil {
		return Result{}, err
	}
```

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+	ctx := context.Background()
+	for _, i := range sampleInputs {
+		var inp Input
+		err := json.Unmarshal([]byte(i), &inp)
+		if err != nil {
+			return err
+		}
```
nit:
```suggestion
		if err := json.Unmarshal([]byte(i), &inp); err != nil {
			return err
		}
```

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+			return err
+		}
+
+		_, err = e.Eval(ctx, inp)
+		if err != nil {
+			return errors.New("policy initialization: issue testing configured policy")
+		}
```
nit:
```suggestion
		if _, err := e.Eval(ctx, inp); err != nil {
			return fmt.Errorf("policy is misconfigured: %v", err)
		}
```

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+
+// NewEngine returns a new policy engine. Or nil if no
+// config is provided.
+func NewEngine(cfg *EngineConfig) (*Engine, error) {
+	if cfg == nil || cfg.FileProvider == nil {
+		return nil, nil
+	}
```
Returning `nil` from a function named `NewEngine` is a bit weird IMO. I'd rather see this function not called if config is empty. Also, IMO, if the config is provided but doesn't have a valid provider set (i.e. FileProvider) that should probably be a failure instead of a silent ignore?

azdagron commented on pkg/common/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+package policy
```
This package seems fairly server-centric, i.e. in the inputs and outputs and tests that are performed. I wonder if some of it belongs in a server policy package. Since we don't have any agent rego plans yet, maybe this should all be defined under server and then the common code refactored out later if/when needed....

azdagron reviewed (commented):
Thanks for opening this, @lumjjb ! I haven't completely reviewed all of the added documentation, but this should be enough to keep you busy while I wrap it up :)

lumjjb commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,387 @@
+
+| API | Authorized if (OR) |
+| ------------- | ----------- |
+| /spire.api.server.svid.v1.SVID/MintX509SVID | local or admin|
```
fixed!

lumjjb commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,387 @@
+listed options. Where the options are:
+- local: local caller through UNIX socket
+- admin: caller has a SPIFFE ID with an "Admin" flag
+- downstream: caller is a downstream SPIFFE ID
```
fixed in commit

lumjjb commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,387 @@
+| req           | The API call request body | { "filter": {} } |
+
+The request (`req`) is the marshalled JSON object from the [SPIRE
+api sdk](https://github.com/spiffe/spire-api-sdk/).
```
Added text for this in table as well as sentence

lumjjb commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,387 @@
+
+## Permissions file (databinding)
+
+The permissoins file consists of a JSON blob which represents the data that is
```
fixed

lumjjb commented on doc/authorization_policy_engine.md:
````diff
@@ -0,0 +1,387 @@
+}
+```
+
+### Default configurations
````
Agreed, added link to file paths instead.

lumjjb commented:
@azdagron all the comments should be addressed.

azdagron commented:
Awesome, thanks @lumjjb. I'll try and get to a review either tonight or tomorrow morning.

lumjjb commented on pkg/server/endpoints/middleware.go:
```diff
@@ -41,56 +42,6 @@ func Middleware(log logrus.FieldLogger, metrics telemetry.Metrics, ds datastore.
-	localOrAdmin := middleware.AuthorizeAnyOf(local, admin)
-	localOrAdminOrAgent := middleware.AuthorizeAnyOf(local, admin, agent)
-
-	return map[string]middleware.Authorizer{
```
Yes - these are covered by https://github.com/spiffe/spire/blob/main/pkg/server/endpoints/endpoints_test.go#L380-L445

Squashed out a few bugs while developing with these tests :)

lumjjb commented on pkg/server/endpoints/endpoints.go:
```diff
@@ -118,6 +120,14 @@ func New(ctx context.Context, c Config) (*Endpoints, error) {
+			return nil, err
+		}
+	}
+
```
changes made.

lumjjb commented on pkg/server/policy/policy.go:
```diff
@@ -0,0 +1,190 @@
+func NewEngineOrDefault(cfg *EngineConfig) (*Engine, error) {
+	if cfg == nil || cfg.FileProvider == nil {
+		return DefaultAuthPolicy()
+	}
```
changes made.

lumjjb commented on pkg/server/policy/policy.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// Eval determines whether access should be allowed on a resource.
+func (e *Engine) Eval(ctx context.Context, input Input) (result Result, err error) {
```
fixed

APTy commented on pkg/server/policy/policy.go:
```diff
@@ -0,0 +1,190 @@
+		return result, err
+	}
+
+	// TODO(tjulian): figure this out
```
Heh, my guess would be that the `Eval()` call returns a slice of [rego.Result](https://pkg.go.dev/github.com/open-policy-agent/opa/rego#Result) objects, each of which contains a slice of ExpressionValues.

So in this implementation, we're only looking at the **first** result and the **first** expression value within that result, which means we're potentially throwing out extra information that could be useful for policy evaluation or logging purposes. Maybe there could be "deny" style rules to be grabbed, or different reasons for "allow", I can't remember exactly.

So it could be worth looking into what the rest of those result/expressions could contain, or leave that as a (more detailed) TODO for later

lumjjb replied:
Ah i see, since then we've introduced a contract between the policy and the authorization component, specifying specific variables that must be set, and only those variables are queried. These are listed [here](https://github.com/lumjjb/spire/blob/policy_opa/doc/authorization_policy_engine.md#rego-policy). I think this addresses mostly the concerns from this TODO, and like you said we can add on to this list later on.

lumjjb commented:
@azdagron Ready for another round!

azdagron commented on pkg/server/policy/validate.go:
```diff
@@ -0,0 +1,144 @@
+
+// validatePolicy runs a few sample inputs with the policy just to make sure/
+// it doesn't throw any errors
+func (e *Engine) validatePolicy() error {
```
This should have the context threaded through from NewEngineFromRego.

lumjjb replied:
Ah! One slipped through the cracks. Fixed!

amartinezfayo commented on pkg/server/server.go:
```diff
@@ -122,7 +123,12 @@ func (s *Server) run(ctx context.Context) (err error) {
-	endpointsServer, err := s.newEndpointsServer(ctx, cat, svidRotator, serverCA, metrics, caManager)
+	policyEngine, err := policy.NewEngineOrDefault(ctx, s.config.PolicyEngineConfig)
+	if err != nil {
+		return err
```
We should format this error to provide an indication of the context, otherwise it may be difficult to guess that the error happened while creating a new authorization policy engine.

lumjjb replied:
fixed

amartinezfayo commented on pkg/server/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+package policy
```
I'm wondering if the directory name `policy` is specific enough, thinking that we may end up having other kind of _policies_ in the future that may have a collision with this. Maybe `authpolicy` sounds better to me. @azdagron Thoughts?

azdagron replied:
authpolicy seems nice :)

amartinezfayo commented on pkg/server/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+
+// NewEngine returns a new policy engine. Or nil if no
+// config is provided.
+func NewEngine(ctx context.Context, cfg *EngineConfig) (*Engine, error) {
```
This function is not exercised in the tests. I think that would be good to have it covered.

lumjjb replied:
Yup - created tests for `NewEngineFromConfigOrDefault`, and made this function private, since it is just a helper. Should encapsulate all the edge cases for `NewEngine`.

amartinezfayo commented on pkg/server/policy/policy_test.go:
```diff
@@ -0,0 +1,262 @@
+			pe, err := policy.NewEngineFromRego(ctx, tt.rego, store)
+			require.Nil(t, err, "failed to create policy engine")
+
+			res, err := pe.Eval(ctxIn, tt.input)
```
Looks like the test cases cover only happy paths when executing `Eval`. I think that it would be good to have coverage of failures also.

lumjjb replied:
Yup - the way its set up, it's designed so that once a policy engine is successfully created, it should by really hard for `Eval` to actually return an error (I tried a lot of things, including divide by zeros....).. it seems like mainly err returned would be because of a system failure.. This is because of the `vallidatePolicy` function catches a lot of the errors that can happen during policy instantiation, so i've added the case where Eval would fail in the instantiation tests here instead:

https://github.com/lumjjb/spire/blob/policy_opa/pkg/server/authpolicy/policy_test.go#L418-L427

amartinezfayo reviewed (commented):
Thank you @lumjjb for this contribution! It's exciting to see this coming!
I have some comments.

lumjjb commented:
Thanks for the review @amartinezfayo ! Got tied up with some stuff but will look at this tomorrow!

lumjjb commented on doc/spire_server.md:
```diff
@@ -68,6 +68,7 @@ This may be useful for templating configuration files, for example across differ
 | `ratelimit`                 | Rate limiting configurations, usually used when the server is behind a load balancer (see below)  |                                                                |
 | `socket_path`               | Path to bind the SPIRE Server API socket to                                                       | /tmp/spire-server/private/api.sock                             |
 | `trust_domain`              | The trust domain that this server belongs to (should be no more than 255 characters)              |                                                                |
+| `policy_engine`             | The policy engine used to determine authorization decision                                        | default SPIRE authorization policy                             |
```
Agreed on the comments here! Will make these changes... only 1 question on one comment..

> server_full.conf needs to be updated with the new configs.

I think @azdagron had another comment from before on this that we may want to default to not setting it - thus using the default policy? This was to prevent replication of the policy in multiple files. Thoughts on the direction with this?

lumjjb replied:
current updated this to look like:

```
auth_policy_engine {
    opa_file_provider {
        ...
    }
}
```

amartinezfayo replied:
>I think @azdagron had another comment from before on this that we may want to default to not setting it - thus using the default policy? This was to prevent replication of the policy in multiple files. Thoughts on the direction with this?

@lumjjb `server_full.conf` is for documentation purposes only. It has all the possible config settings with comments so users can look at the different configurables in SPIRE. It doesn't look like the comment from @azdagron applies to that, right?

lumjjb replied:
ah i see - yea i got mixed up. Yea that works.

lumjjb commented on doc/spire_server.md:
```diff
@@ -84,6 +85,17 @@ This may be useful for templating configuration files, for example across differ
 
+| policy_engine               | Description                    | Default        |
+|:----------------------------|--------------------------------|----------------|
+| `file_provider`             | File to retrieve OPA rego policy for authorization. |      |
```
I think it's possible that this may also point to an external OPA endpoint, so there's that possibility that instead of a file it would point to a server with some connection configuration.

lumjjb replied:
not sure about other alternatives to OPA, i think there is a realm of possibility but not sure what that would be today... Am in favor of adding an OPA specific config in this case.

lumjjb replied:
renamed to `opa_file_provider`

lumjjb commented on pkg/server/policy/policy.go:
```diff
@@ -0,0 +1,170 @@
+	// Test policy with some simple calls to ensure that the
+	// policy can be evaluated properly.
+	if err := e.validatePolicy(ctx); err != nil {
+		return nil, err
```
fixed

lumjjb commented on pkg/server/api/middleware/authorization_opa.go:
```diff
@@ -0,0 +1,130 @@
+
+	if res.AllowIfAgent {
+		ctx, isAgent, err := isAgent(ctx, agentAuthorizer)
+		if err == nil && isAgent {
```
Yea good catch.

Just to provide some context for when I was writing this.. 

`isAgent` may return an error if it can't figure out if its an agent, but if an authorization policy says please allow "if agent" or "if local" then this call to `isAgent` would error.. so the order of operations is what really "saves" it here. 

Maybe... the way to address this is changing `isAgent` such that it will not return an error if caller does not have an `x509`?

lumjjb commented:
@amartinezfayo I think I should've addressed all of the comments.. 🤞 . Thanks for the great comments!

amartinezfayo commented on doc/spire_server.md:
```diff
@@ -68,6 +68,7 @@ This may be useful for templating configuration files, for example across differ
 | `ratelimit`                 | Rate limiting configurations, usually used when the server is behind a load balancer (see below)  |                                                                |
 | `socket_path`               | Path to bind the SPIRE Server API socket to                                                       | /tmp/spire-server/private/api.sock                             |
 | `trust_domain`              | The trust domain that this server belongs to (should be no more than 255 characters)              |                                                                |
+| `auth_policy_engine`        | The [auth policy engine](doc/authorization_policy_engine.md) used for authorization decision      | default SPIRE authorization policy                             |
```
This document is already in `/doc` so a relative reference to `doc` will link to /doc/doc.

```suggestion
| `auth_policy_engine`        | The [auth policy engine](/doc/authorization_policy_engine.md) used for authorization decisions      | default SPIRE authorization policy                             |
```

amartinezfayo reviewed (commented):
Thanks @lumjjb! One more small comment.

lumjjb commented:
Committed the suggestion! let me know if you'd like me to rebase the PR

Benardi commented:
Minor detail, but the documentation at `doc/authorization_policy_engine.md` still points to `pkg/server/policy/policy.rego` and  `pkg/server/policy/permissions.json` instead of  `pkg/server/authpolicy/policy.rego` and `pkg/server/authpolicy/permissions.json`

amoore877 commented on doc/authorization_policy_engine.md:
````diff
@@ -0,0 +1,224 @@
+}
+```
+This sets `check_entry_create_admin_flag` to true if the full method is not for
+entry creation or if it is, that there are no entires that contain the admin
````
typo: `entires` in a few places

amoore877 commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,224 @@
+
+This section contains examples of how the authorization policy can be extended.
+
+## Example 1: Entry creation namespacing restrictions
```
`Entry creation namespacing restrictions` is a great example to have; I think for the complexity of authorization policies it would be helpful to also add examples for:
- a "subdepartment" permissions split. Say `schedulers/finance` is able to create `/finance` and also `schedulers/finance/EMEA` is able to create `/finance/EMEA`.
- ^Are we able to also create such that `schedulers/finance` can still create `/finance` but is excluded from creating `/finance/EMEA`, or is that not possible?
- having a delete or update example may also be helpful, just so examples aren't restricted to single API

lumjjb replied:
good idea - let me add some other examples that cover these.. 

>^Are we able to also create such that schedulers/finance can still create /finance but is excluded from creating /finance/EMEA, or is that not possible?

it should be possible.. it would just a bit more complicated to express. It may be a difficult example to follow, let me mock up some examples so we can take a look

lumjjb replied:
Doc updated with some more examples!

evan2645 commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,224 @@
+  caller is a local UNIX socket call
+- `allow_if_admin`: a boolean that if true, will authorize the call only if the
+  caller is a SPIFFE ID with the Admin flag set
+- `allow_if_downstream`: a boolean that if true, will authorize the call 
```
I think we're missing `allow_if_agent` ?

lumjjb replied:
good catch - fixed

evan2645 commented on doc/authorization_policy_engine.md:
````diff
@@ -0,0 +1,224 @@
+}
+```
+
+The fields of the result are the following:
````
What happens if more than one of the fields is true? What if both `allow` and `allow_if_agent` are true?

lumjjb replied:
ah yea - forgot to include this, copying this from the conversation in the issue.

evan2645 reviewed (commented):
This is great, thank you so much @lumjjb! We're way overdue for this

I had a bunch of thoughts about the shape and naming of the configs, but I'm happy to circle back around in a follow-up PR with a proposal so as not to block on it here. I think the only _real_ thing is the docs question about edge case on the response object. Also missing the `allow_if_agent` bit in the docs ... other than that, lgtm

lumjjb commented on conf/server/server_full.conf:
```diff
@@ -102,6 +102,19 @@ server {
+    #         # Path to the policy rego file
+    #         policy_path = "./conf/server/policy.rego"
+    #         # Path to the permissions JSON databindings
+    #         permissions_path = "./conf/server/permissions.json"
```
Yea - I agree very much on this - initial use of this was only for permissions, or at least the common usecase is to encapsulate permissions. Yea I think i will comb through and change it to `policy_data_path`.

lumjjb commented on conf/server/server_full.conf:
```diff
@@ -102,6 +102,19 @@ server {
+    # auth_policy_engine {
+    #     # Indicates that OPA provided through local files be used to 
+    #     # instantiate the auth policy engine
+    #     opa_file_provider {
```
>The fact that we are using OPA internally feels like a little bit of an implementation detail... is it "just" rego evaluation, or are there OPA-specific things here? Would it make sense for the name of this provider to be something about rego instead?

I think rego is really only tied to only OPA right now, the evaluation semantics are still all OPA specific, i.e. handling of errors, execution order, etc. So I think in this case "opa" may be a more direct translation of what's happening since it encapsulates both the static and dynamic semantics. 

>It would be nice if the name was a bit shorter too ... the provider suffix is probably redundant, perhaps something like rego_file or opa_file
>I imagine that if we had rego_file, then we may in the future have something like local_opa etc?

Yea, I really like `local_opa` even now, since most likely the next extension to this is `remote_opa` where policy can be defined in a remote server. What do you think about this?

lumjjb commented on conf/server/server_full.conf:
```diff
@@ -102,6 +102,19 @@ server {
 
+    # auth_policy_engine: The auth policy engine used for authorization decision
+    # For more details, refer to doc/authorization_policy_engine.md
+    # auth_policy_engine {
```
Addressing the other comment together with this... Maybe in that case, perhaps we can just call it

Hmm... only thing that comes to mind here is around when one OPA policy engine fails to evaluate, then there's a fallback policy engine. i.e. remote OPA policy server is down, default back to local OPA policy. 

`auth_opa_policy_engine` and properties of it would be `local` 

```
auth_opa_policy_engine = {
  local {
    policy_path = xyz
    policy_data_path = xyz
  }
}
```

Benardi commented on doc/authorization_policy_engine.md:
````diff
@@ -0,0 +1,235 @@
+```
+# Any allow check
+allow = true {
+    check_entry_create_admin_flag
````
Shouldn't it be `check_entry_create_namespace`?

I'm met with the following error if I use this line as it stands, which I surmise might arise from `check_entry_create_namespace` not being defined:

```bash
O[0000] Fatal run error                               error="unable to obtain authpolicy engine: 1 error occurred: spire.rego:97: rego_unsafe_var_error: var check_entry_create_admin_flag is unsafe"
ERRO[0000] Server crashed                                error="unable to obtain authpolicy engine: 1 error occurred: spire.rego:97: rego_unsafe_var_error: var check_entry_create_admin_flag is unsafe"
```

lumjjb replied:
yes - you are right, i copy-pasted and forgot to change 😓

Benardi commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,235 @@
+
+This section contains examples of how the authorization policy can be extended.
+
+## Example 1: Entry creation namespacing restrictions
```
I've tried to follow this example. I've built the server and agent using `make build` after editing `policy.rego` and `permissions.json`. 

As a test, I've done the following:
*  I've created an agent with the following SPIFFE ID:  **spiffe://example.org/schedulers/finance**
* I then created the following entry
  + SPIFFE ID **//example.org/schedulers/hr**
   + Parent ID **spiffe://example.org/schedulers/finance**

I managed to create the entry and even receive the corresponding SVID from the Workload API (via [ghostunnel](https://github.com/ghostunnel/ghostunnel)).
Shouldn't this have failed?

lumjjb replied:
Hey Benardi! Let's chat on slack see whether we can figure out what the issue is - or whether my OPA code has a bug

Benardi replied:
Sure thing.

Benardi replied:
So far, we've established that as I was using `spire-server entry create`, the call to BatchCreateEntry API must have been marked as local, and thus namespacing was bypassed.

I still have to surmise a way to make a non-local call to the BatchCreateEntry API to discern whether the namespacing will be enforced.

PhilCorney replied:
Hey Benardi,
Did you find a way to test calling the Entry API through Ghostunnel?
I'm looking to test something simliar. I've got the [tunnel](https://github.com/ghostunnel/ghostunnel/tree/master/docs/spiffe-workload-api-demo/unix) working was going to configure the front-end-agent to call the server tcp connection, but unsure how to form the request.
Many thanks
Phil

lumjjb commented:
@evan2645 since there's a couple of comments around the config, going to summarize the different comments and replies in 1 place for easier discussion, I am in agreement with all of them and want to propose a new config format for discussion. Let me know what you think about the new proposed config/comments and I can make the changes.


- If we are accommodating future support for other authorization engines, would a  plugin would be more appropriate - possibly in the future?
  - Yep agreed, if we get more, agree that plugin would be more appropriate. No current foreseeable plans to have other authorization engines for now.
- can there be more than one policy engine set? The shape of this config makes it feel that way. It's hard for me to imagine the uses cases for multiple policy engines, and what the expected behavior might be if we get different answers etc ... should we try to shape the config so that it's clear that there can be only one?
  - Yea - agree on having multiple policy engines isnt that useful. The only scenario that i see this useful for example is in OPA if we have a remote and a local option, then we can. set some behavior where if remote fails to query, fall back to a fail secure local OPA policy.
- The fact that we are using OPA internally feels like a little bit of an implementation detail... is it "just" rego evaluation, or are there OPA-specific things here? 
  - Rego is pretty much tied to OPA right now, but OPA encapsulates the runtime semantics as well (i.e. how the rego is parsed and how errors are handled), so i think would be more precise to classify it as "opa". (and the fact that it takes in a data binding as well)
- It would be nice if the name was a bit shorter too ... the provider suffix is probably redundant, perhaps something like rego_file or opa_file.I imagine that if we had rego_file, then we may in the future have something like local_opa etc?
  - Changing this to local
- data bindings could be many things right? Not just permissions?
  - Yup - changed to policy_data_path

## ** Proposed config ** - that hopefully addresses all the comments
```
auth_opa_policy_engine = {
  local {
    rego_path = xyz
    policy_data_path = xyz
  }
}
```

lumjjb commented:
@amoore877 @azdagron @evan2645 @amartinezfayo Ok! I think i've addressed all the comments! 🤞

azdagron commented:
The bit I was worried about (in terms of behavior regression) has been addressed. I'll leave it up to @evan2645 to comment on the shape of the configuration :)

amoore877 commented on doc/authorization_policy_engine.md:
```diff
@@ -0,0 +1,374 @@
+
+It is important when implementing custom policies that one understands the 
+evaluation semantics and details of OPA rego. An example of subtleties of OPA
+rego policy is the evaluation of a variable is taken as an logical OR of all 
```
nit (please don't bother if no other changes being made): `as an logical OR` -> `as a logical OR`

amoore877 commented on doc/authorization_policy_engine.md:
````diff
@@ -0,0 +1,374 @@
+}
+```
+
+## Example 1b: Sub-department namespacing with exlcusions
````
Thanks for adding!

amoore877 commented on pkg/server/api/middleware/authorization.go:
```diff
@@ -59,18 +51,18 @@ func (m *authorizationMiddleware) Preprocess(ctx context.Context, methodName str
+	}
+
+	deniedErr := status.Errorf(codes.PermissionDenied, "Authorization denied for method %s", methodName)
+	rpccontext.Logger(ctx).WithError(deniedErr).Error("Failed to authenticate caller")
```
"Failed to authenticate caller" is handled higher up in the function, where we got some error in the opa auth. I think this error log should be "Failed to authorize caller", or maybe better "Caller is unauthorized" since it's not really a failure of the service but instead an explicit rejection.

lumjjb replied:
i'm not sure if i understood your comment correctly. i've updated it in https://github.com/spiffe/spire/pull/2416/commits/bdf9ad4736eec98efa955e2d778b3e2b6da7a348, if you can take a quick look.

`Failed to authenticate caller` is what is being passed back today if the authentication middleware denies.

amoore877 replied:
The new wording addresses my comment.

The previous logic had a general error for authorization, which could mean either the auth itself failed (could be authn or authz failure) or there was no "error" but instead the caller is rejected but authz. So originally it may have somewhat made sense to bundle all that into a single error return.

With this change, we now explicitly know if there was an error, or if there was no error and the call is authorized or not. An error now is actually a real error, and the true/false boolean is an explicit authz accept or deny return. So for the flag check, there is no "failure" error to log, but instead "you're not authorized". Does that make sense?

lumjjb replied:
Ah yes! Gotcha, thanks for providing more detail!

amoore877 commented on pkg/server/api/middleware/authorization.go:
```diff
@@ -59,18 +51,18 @@ func (m *authorizationMiddleware) Preprocess(ctx context.Context, methodName str
+		return ctx, nil
+	}
+
+	deniedErr := status.Errorf(codes.PermissionDenied, "Authorization denied for method %s", methodName)
```
This entire effort is a fantastic security addition. Has there been discussion for future work on enabling triage of auth failures? In particular, I am thinking of the Batch APIs. On rejection, the caller (and SPIRE operators) only seem to know that at least one of the entries in the Batch was denied, but not which one. Depending on complexity of policies being set, it could be a bit of work to determine denial reason. If the caller was a workload, it is also cumbersome to get the entire list of entries printed out.

lumjjb replied:
yea, that's a great idea!!! :). Having error messages being propagated from the OPA engine would be super useful. I'm still unsure what the best way to do this is with OPA. Let me ask some of the OPA folks to see if they have anything that can do this in the pipeline or if they have some ideas of how this can be done today.

evan2645 replied:
In general, reason of denial is a pretty big deal, and as complexity grows this gets more important as @amoore877 is noting here. This is another general issue with adopting opa/rego in this fashion, so if there's not an existing solution I imagine that there is a community need for one.

lumjjb replied:
There's an example of it here https://play.openpolicyagent.org/p/giimmNjfrj

it does look like the policies need to be instantiated a certain way - like a deny list rather than an allow list , will have to look into this a bit more.

amoore877 commented:
Awesome work :)

evan2645 reviewed (approved):
thanks for this contribution @lumjjb ... it is hugely impactful!

There's a merge conflict, but the comments I raised previously have been fully addressed. Can't wait to see what people cook up using this :P
